README.md
v0.1 (c)2016 by @errecepe [CC BY-NC-SA] 3.0 license
29/10/2016
-----------------------------------------------------------------------------------------
Description

Get's remote nagios host and service status value. Useful for Nagios Distributed Environment without installing extras on Main Nagios server
-----------------------------------------------------------------------------------------
Basic help

usage: 
	check_remotenagios.pl [remote_nagios_host] [nagios_user] [password] [host] [service] 

Where: 
	[remote_nagios_host] = ip address of remote nagios server to check 
	[nagios_user] = nagios user 
	[password] = nagios user password 
	[host] = name of remote host in nagios 
	[service] = name of remote service in nagios

For example the command: 
	./check_remotenagios.pl 192.168.1.50 nagiosadmin pa55w0rd localhost SSH 
	Will check service 'SSH' on host 'localhost' on remote server nagios with IP:192.168.1.50 using nagiosadmin with pa55w0rd as password

Others params:
	--version	show version
	--help		show help

-----------------------------------------------------------------------------------------
Instalation and configuration:

Step 1:
  In 'main Nagios server' copy in /usr/local/nagios/libexec/ the script file check_remotenagios.pl, and chmod 777 check_remotenagios.pl

Step 2:
  In 'main Nagios server' add a command called: check_remotenagios
  with command: $USER1$/check_remotenagios.pl $ARG1$ $ARG2$ $ARG3$ "$ARG4$" "$ARG5$"

Step 3:
  apply changes, and restart nagios service after verify configuration (usually /usr/local/nagios/bin/nagios -v   /usr/local/nagios/etc/nagios.cfg)

Done! It's ready for use!
The script it's developed using Perl and curl (these tools are usually currently installed in all nagios servers), not need 
extra configurations or installation on main Nagios server.
-----------------------------------------------------------------------------------------
Using command:

Now you can add services using the new command check_remotenagios (or the name you used in Step 2) in your 'main Nagios server'
and configure for get values from a remote nagios server. Obviously the checked host and service must be defined on remote Nagios
server. 

The script will use just curl (need port 80 opened, can check it in'main Nagios server' executing "nmap REMOTE_NAGIOS_IP -p 80") to
connect via http to remote nagios. 

The params in nagios are:

$ARG1$ its ip address of remote nagios server to check [remote_nagios_host]
$ARG2$ its nagios user [nagios_user]
$ARG3$ its nagios user password [password]
$ARG4$ its name of remote host in nagios [host]
$ARG5$ its name of remote service in nagios [service]

You will get the state, and performance data from remote nagios server, so under Nagios XI you will get data for performance graphs.

This script can be used for distribute checks in a complex production environment. Or if you need to use checks using a big cpu load,
can use an external server for distribute cpu load on 'main nagios server' easily.

Tested with Remote nagios core 4 (4.1.1) version, from Nagios XI 5 (5.2.5) as main Nagios server.
Supposed to work in any nagios xi or core version, but ONLY with remote core version. 
-----------------------------------------------------------------------------------------
22:50 01/11/2016 Edit:

Can be user with Nagios XI remote version following these steps:

1	edit /etc/httpd/conf.d/nagios.conf
	modify these lines
		#   AuthName "Nagios Core"
		#   AuthType Basic
		#   AuthUserFile /usr/local/nagiosxi/etc/htpasswd.users
		#   Require valid-user
	CAUTION:We will give free access to NAGIOSXISERVER\nagios with this action!!!

2	in Nagios XI web, Config, CCM, Nagios Core CGI Config
	modify line
		use_authentication=0

3	from ssh restart httpd and nagios services

Now we will check remote Nagios XI servers too!	

Use it at your own risk!


v0.1/changelog.txt
v0.1 0:21 28/10/2016

- script base development

v0.1 3:21 29/10/2016

- fixed host and service params passed as "With Spaces" now are working ok.
  commands in nagios main server must be defined as: $USER1$/check_remotenagios.pl $ARG1$ $ARG2$ $ARG3$ "$ARG4$" "$ARG5$"

- included in source a modification for use the script under windows os. Can be used from cmd, powershell or other scripts, to get info from      remote nagios servers. Requires curl command avalaible in PATH installed on windows host.

- modified script to return as Status Information the original one.

- final code optimization

v0.1/check_remotenagios.pl
#!/usr/bin/perl
# Coded by: @errecepe
# Start Date: 2016/10/26
# End Date v0.1 2016/10/29
#
# usage: check_remotenagios.pl nagioshost nagioshost_user nagioshost_password host service
#

#add command on nagios as
# 
#	name:		check_command
#	command:	$USER1$/check_remotenagios.pl $ARG1$ $ARG2$ $ARG3$ "$ARG4$" "$ARG5$"
#

use HTML::Parser(); 

$STATE_OK = 0;
$STATE_WARNING = 1; 
$STATE_CRITICAL = 2;
$STATE_ERROR = 255;

#basic params

if (@ARGV[0] eq "--help"){showhelp();exit $STATE_ERROR;}
if (@ARGV[0] eq "--version"){showversion();exit $STATE_ERROR;}

#if less than 5 parameters, show help and exit
if ($#ARGV + 1 < 5) {
	showhelp();
	exit $STATE_ERROR;
}

$NAGIOSREMOTEHOST=@ARGV[0];
$NAGIOSUSER=@ARGV[1];
$NAGIOSPASSWORD=@ARGV[2];
$HOST=@ARGV[3];
$HOST=join("%20", split(" ", $HOST)); #replace spaces for %20 for correct params url syntax calling
$SERVICE=@ARGV[4];
$SERVICE=join("%20", split(" ", $SERVICE)); #replace spaces for %20 for correct params url syntax calling


# curl remote nagios
$RAWHTML=`echo 'user=$NAGIOSUSER:$NAGIOSPASSWORD'|curl -s -K - "http://$NAGIOSREMOTEHOST/nagios/cgi-bin/extinfo.cgi?type=2&host=$HOST&service=$SERVICE"`;
#comment previous line and uncomment next for execute this script under Windows  
#$RAWHTML=`echo user=$NAGIOSUSER:$NAGIOSPASSWORD|curl -s -K - "http://$NAGIOSREMOTEHOST/nagios/cgi-bin/extinfo.cgi?type=2&host=$HOST&service=$SERVICE"`;
#


#get state service from div in HTML source from check remote
$estaok=index($RAWHTML,'serviceOK');
if ($estaok==-1){
		$estawarning=index($RAWHTML,'serviceWARNING');
		if ($estawarning==-1){
		$estaerror=index($RAWHTML,'serviceCRITICAL');
			if ($estaerror==-1){ $resSTATE = $STATE_ERROR ; }else{ $resSTATE = $STATE_CRITICAL ;}
		}
		else
		{
		$resSTATE=$STATE_WARNING;
		}
	}
	else
	{
		$resSTATE=$STATE_OK;
	}
#end 	

#search state info
$posinfo=index($RAWHTML,'Status Information:')+44	; #44 caracteres contando el codigo html que devuelve nagios via web, podría variar segun cliente, funciona ok con core4
$posinfoend=index($RAWHTML,'</TD>',$posinfo)	; #cojo el texto entre $posinfo y el siguiente </td>
$STATUS_INFO = substr($RAWHTML,$posinfo, $posinfoend-$posinfo);
#end search state info

#search performance data
$ppdatainfo=index($RAWHTML,'Performance Data:')+42	; #44 caracteres contando el codigo html que devuelve nagios via web, podría variar segun cliente, funciona ok con core4
$ppdataend=index($RAWHTML,'</TD>',$ppdatainfo)	; #cojo el texto entre $posinfo y el siguiente </td>
$PP_DATA = substr($RAWHTML,$ppdatainfo, $ppdataend-$ppdatainfo-51); #no tengo claro de donde sale el 51
#end search performance data

#finally show error or correct info, and return a correct state code for nagios, and exit
if ( $resSTATE == $STATE_ERROR ) 
	{
	print "UNKNOWN ERROR (nagios server incorrect, credentials incorrect, or host+service doesnt exist)\n"; 
	exit $STATE_ERROR;
	}
	else
	{
	print "$STATUS_INFO |$PP_DATA\n"; 
	exit $resSTATE;
	}

#Help command
sub showhelp()
{
	showversion();
	print "\nGet's remote nagios host and service status value. Useful for Nagios Distributed Environment without installing extras on Main Nagios server\n";
	print "\n";
	print "usage:\n";
	print "  check_remotenagios.pl [remote_nagios_host] [nagios_user] [password] [host] [service]\n";
	print "Where:\n";
	print "  [remote_nagios_host]	= ip address of remote nagios server to check\n";
	print "  [nagios_user]		= nagios user\n";
	print "  [password]		= nagios user password\n";
	print "  [host]		= name of remote host in nagios\n";
	print "  [service]		= name of remote service in nagios\n";
	print "\nFor example the command:\n";
	print "  ./check_remotenagios.pl 192.168.1.50 nagiosadmin pa55w0rd localhost SSH\n";
	print "Will check service 'SSH' on host 'localhost' on remote server nagios with IP:192.168.1.50 using nagiosadmin with pa55w0rd as password\n";
}

#Version command
sub showversion()
{
	print "check_remotenagios.pl v0.1  by \@errecepe more info at https://github.com/errecepe/check_remotenagios.pl 2016/10/28 \n";
}